### ISSUE TITLE
Gatsby proxy DELETE request strips request body

### ISSUE THREAD

**Issue by kylegoetz** [gatsbyjs/gatsby#24165]
## Description

When I have a proxy configured with Gatsby during development and make a DELETE request with a request body, Gatsby forwards the request but strips out the request body. (**Edit** Actually now I'm not so sure it actually forwards the request rather than hanging without forwarding)

### Steps to reproduce

Using `axios` as the request library, and assuming Gatsby is set up to run on port `3000` and proxies a backend running on, say, port 8000:

    curl -X DELETE --data '{"foo":"bar"}' http://localhost:3000

What gets sent to the server on port 8000 is a request with an empty request body. However, if you

    curl -X DELETE --data '{"foo":"bar"}' http://localhost:8000

directly hitting the backend, the request body is present.

**Minimal reproducible example** https://github.com/kylegoetz/gatsby-delete-bug

1. `npm run test` to instantiate the HTTP server (port 10000) and Gatsby (port 8000)
2. (see repo readme) in another terminal, use curl to POST data to Gatsby (which proxies the HTTP server during development). The response should include the data (and the other CLI where you ran `npm run test` should log the POST request body.
3. (see repo readme) now use the same curl command but with DELETE instead of POST. Gatsby does not forward the request and instead the query just hangs.

According to the HTTP spec, DELETE can have a request body, so the query is as expected. (Indeed, if you directly DELETE request the server on port 10000, it correctly resolves as the POST does)

### Expected result

What should happen is the request body remains. The HTTP spec states that DELETE requests can have a body. (I believe this used to not be the case a long time ago, but certainly for a few years it's true because the software I'm debugging right now is about five years old and being ported to a Gatsby project).

### Actual result

See above. Empty request body.

### Environment

Run `gatsby info --clipboard` in your project directory and paste the output here.

```
 System:
    OS: macOS Mojave 10.14.6
    CPU: (4) x64 Intel(R) Core(TM) i5-4258U CPU @ 2.40GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 13.9.0 - /usr/local/bin/node
    Yarn: 1.22.4 - /usr/local/bin/yarn
    npm: 6.13.7 - /usr/local/bin/npm
  Languages:
    Python: 2.7.15 - /usr/local/bin/python
  Browsers:
    Chrome: 81.0.4044.138
    Firefox: 76.0.1
    Safari: 13.1
  npmPackages:
    gatsby: ^2.3.0 => 2.20.12
    gatsby-image: ^2.0.0 => 2.3.1
    gatsby-plugin-layout: ^1.0.0 => 1.2.1
    gatsby-plugin-manifest: ^2.0.0 => 2.3.3
    gatsby-plugin-notify: ^1.0.0 => 1.0.0
    gatsby-plugin-nprogress: ^2.0.0 => 2.2.1
    gatsby-plugin-offline: ^2.0.0 => 2.2.10
    gatsby-plugin-polyfill-io: ^1.0.0 => 1.1.0
    gatsby-plugin-postcss: ^2.0.0 => 2.2.1
    gatsby-plugin-preact: ^3.1.8 => 3.2.1
    gatsby-plugin-preconnect: ^1.0.5 => 1.0.5
    gatsby-plugin-react-css-modules: ^2.0.0 => 2.2.1
    gatsby-plugin-react-helmet: ^3.0.0 => 3.2.1
    gatsby-plugin-robots-txt: ^1.4.0 => 1.5.0
    gatsby-plugin-sass: ^2.2.1 => 2.2.1
    gatsby-plugin-sharp: ^2.0.0 => 2.5.4
    gatsby-plugin-sitemap: ^2.0.0 => 2.3.1
    gatsby-plugin-svgr: ^2.0.0 => 2.0.2
    gatsby-plugin-typescript: ^2.3.1 => 2.3.1
    gatsby-plugin-webpack-bundle-analyzer: ^1.0.0 => 1.0.5
    gatsby-transformer-sharp: ^2.1.0 => 2.4.4
  npmGlobalPackages:
    gatsby-cli: 2.11.5
```

**Comment by blainekasten** [comment]
Hey @kylegoetz!

Thanks for reporting the above issue, and sorry for the trouble you're running into. I've never heard of anyone needing to send DELETE requests to the gatsby development server. Can you explain a bit more what you are trying to achieve?

Under the hood we are running a light express API and using webpack-hot-middleware. I don't think our setup has anything specific to it, but it doesn't seem to be designed to support any DELETE requests. What are you trying to achieve?

**Comment by kylegoetz** [comment]
I have a restful api I need to interact with during development. After build, the static site and restful api are both served by Nginx/apache.

I set the `proxy` config in `gatsby-config.js` as per the Gatsby documentation to proxy the requests as I develop the UI.

There are a number of http actions you can do to interact with a server. Get, post, put, patch, delete, etc.

In restful architecture, get fetches data, post creates it, put/patch are for updating data, and delete is for deleting it.

So if I want to make an API request to delete a user's project number 57825, for example, in development I make a DELETE request to `http://localhost:port/api/v1/projects/57825`.

But for a few requests I have to pass additional data on the DELETE request. That goes in the request body.

I followed these instructions on Gatsby's site to set up the proxying: https://www.gatsbyjs.org/docs/api-proxy/#proxying-api-requests-in-development

Let me know if you need more info. If I can get away from my toddler for enough time today, I want to dig into where this is happening. I know express doesn't have any problem with delete requests, as that is what our backend uses.

**Comment by kylegoetz** [comment]
For my own tracking purposes, here is the code where the proxy is set up in Gatsby https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby/src/commands/develop.ts

**Comment by kylegoetz** [comment]
Found the problem. The proxy server created by Gatsby isn't enabling cors, so the pre-flight requests aren't forwarded. This means GET, POST, and OPTIONS work, but PATCH, PUT, DELETE will not.

The Express docs explain it here, that you have to pass cors to all the middleware. I see it passed at some point in another section of the code, but not for the proxy part. https://expressjs.com/en/resources/middleware/cors.html

I'm reading the docs for how to contribute to Gatsby and will submit a PR.

**Comment by kylegoetz** [comment]
OK I have the fix implemented by following https://www.gatsbyjs.org/contributing/setting-up-your-local-dev-environment/

I want to add tests even though I'm sure it works since I've already used my new version and it behaves as expected. Where should I put the tests?

I suppose it needs an end-to-end test on Gatsby's dev proxy functionality, so it should spin up a dummy server to be what is proxied, make DELETE, PUT, and PATCH requests with request bodies, have the dummy server response contain the request body as its own response body, and the test will verify the two bodies are the same.

My guess is somewhere in `e2e-tests/development-runtime/src`. Not sure how to do this, though.

I followed the instructions but do not understand what the instructions in "Gatsby functional changes" require of me. We just need the Gatsby proxy to implement cors, which it should to be a dev proxy (otherwise it only proxies GET/POST rather than the full suite of HTTP verbs), but it currently does not.

I'll go ahead and submit the PR minus any additional tests. I don't currently see any tests specifically for the Gatsby proxy functionality, so maybe it's not needed for the PR anyway.

:)
